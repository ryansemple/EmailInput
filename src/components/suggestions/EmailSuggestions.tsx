import React, { useState, useEffect } from "react";
import EmailSuggestion from "./EmailSuggestion";
import { emailAppearsToBeValid } from "../../utility/Language";
import EmailSuggestions from "../../types/EmailSuggestions";

interface EmailSuggestionsProps {
	email: string,
	setEmail: (email: string) => void,
	setEmailIsValid: (emailIsValid: boolean) => void,
	setEmailMessage: (emailMessage: string) => void
}

/**
 * Shows valid email suggestions based on the user's inputted email.
 * @param {EmailSuggestionsProps} props
 */
const EmailSuggestionsComponent = (props: EmailSuggestionsProps) => 
{	
	const { email, setEmail, setEmailIsValid, setEmailMessage } = props;
	const [currentEmailSuggestions, setCurrentEmailSuggestions] = 
		useState([""]);
	const [showEmailSuggestions, setShowEmailSuggestions] = 
		useState(false);
	const [defaultEmailSuggestions] = useState(new EmailSuggestions());

	
	useEffect(() => 
	{
		setCurrentEmailSuggestions
		(
			defaultEmailSuggestions.returnValidEmailSuggestions(email)
		);

		if (email)
		{
			setShowEmailSuggestions(true);
		} 
		else 
		{
			setShowEmailSuggestions(false);
		}
	}, 
	[
		email,
		setShowEmailSuggestions,
		defaultEmailSuggestions
	]);

	/**
	 * when the suggestion is clicked, the email in the form input
	 * will be changed to the suggestion and the user will be able
	 * to submit the email for verification.
	 * @param {string} suggestedEmail - the clicked suggested email.
	 */
	const emailSuggestionOnClick = (suggestedEmail: string): void => 
	{
		setEmail(suggestedEmail);
		setEmailIsValid(true);
		setEmailMessage(emailAppearsToBeValid);
	}
	
	return (
		<>
			{showEmailSuggestions && 
			<ul className="EmailSuggestions full_width">
				{currentEmailSuggestions.map((emailSuggestion: string) => 
					<EmailSuggestion
						key={emailSuggestion}
						emailSuggestion={emailSuggestion}
						onClick={emailSuggestionOnClick}
					/>
				)}
			</ul>
			}
		</>
	)
}

export default EmailSuggestionsComponent;