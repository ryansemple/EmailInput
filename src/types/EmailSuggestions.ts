import EmailSuggestion from "./EmailSuggestion";
import { emailDomains, domainExtensions } from "../utility/Email";
import { atCharacter } from "../utility/String";

interface EmailSuggestionsInterface {
	emailSuggestions: EmailSuggestion[],
	returnValidEmailSuggestions: (userEnteredEmail: string) => string[]
}

class EmailSuggestions implements EmailSuggestionsInterface {
	public readonly emailSuggestions: EmailSuggestion[];
	
	constructor()
	{ 
		let emailSuggestions: EmailSuggestion[] = [];
		
		for (let i: number = 0; i < emailDomains.length; i++)
		{
			const emailDomain: string = emailDomains[i];

			for (let i2: number = 0; i2 < domainExtensions.length; i2++)
			{
				const domainExtension: string = domainExtensions[i2];
				
				emailSuggestions.push
				(
					new EmailSuggestion(emailDomain, domainExtension)
				);
			}
		}

		this.emailSuggestions = emailSuggestions;
	}

	public returnValidEmailSuggestions = (userEnteredEmail: string): string[] =>
	{
		if (!userEnteredEmail.includes(atCharacter)) 
		{
			return this.returnAllEmailSuggestions(userEnteredEmail);
		}
		else 
		{
			const emailSplitByAtCharacter: string[] = 
				userEnteredEmail.split(atCharacter);
			const emailCharactersAfterAtCharacter: string = 
				emailSplitByAtCharacter[1];

			if (emailCharactersAfterAtCharacter === "") 
			{
				return this.returnAllEmailSuggestions(userEnteredEmail);
			}
			else
			{
				return this.returnFitleredEmailSuggestions(
					userEnteredEmail, 
					emailCharactersAfterAtCharacter
				);
			}
		}
	}

	private returnAllEmailSuggestions = (userEnteredEmail: string): string[] =>
	{
		let validEmailSuggestions: string[] = [];
		
		for (let i: number = 0; i < this.emailSuggestions.length; i++)
		{
			const currentEmailSuggestion: EmailSuggestion = this.emailSuggestions[i];
			const validEmailSuggestion = currentEmailSuggestion.returnFullEmail(
				userEnteredEmail
			);
			validEmailSuggestions.push(validEmailSuggestion);
		}

		return validEmailSuggestions;
	}

	private returnFitleredEmailSuggestions =
	(userEnteredEmail: string, userEnteredEmailAfterAtCharacter: string) : string[] => 
	{
		let validEmailSuggestions: string[] = [];

		for (let i: number = 0; i < this.emailSuggestions.length; i++)
		{
			const currentEmailSuggestion: EmailSuggestion = this.emailSuggestions[i];
			const currentEmailSuggestionAfterAtCharacter: string =
				currentEmailSuggestion.returnEmailAfterAtCharacter();
			const regexToMatch: RegExp = new RegExp
			(
				`^${userEnteredEmailAfterAtCharacter}`
			);

			const userEmailMatchesSuggestedEmailExactly = 
				userEnteredEmailAfterAtCharacter ===
				currentEmailSuggestionAfterAtCharacter;
			
			if(userEmailMatchesSuggestedEmailExactly) 
			{
				continue;
			}

			if (currentEmailSuggestionAfterAtCharacter.match(regexToMatch))
			{
				const validEmailSuggestion = currentEmailSuggestion.returnFullEmail(
					userEnteredEmail
				);

				validEmailSuggestions.push(validEmailSuggestion);
			}
		}

		return validEmailSuggestions;
	}
}

export default EmailSuggestions;