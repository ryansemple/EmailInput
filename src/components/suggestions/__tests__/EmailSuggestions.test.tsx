import React from "react";
import { render } from "@testing-library/react";
import EmailSuggestions from "../EmailSuggestions";

test("Renders EmailSuggestions component correctly.", () => {
	const { container } = render
	(
		<EmailSuggestions
			email="ryan@gmail.com"
			setEmail={() => {}}
			setEmailIsValid={() => {}}
			setEmailValidationMessage={() => {}}
		/>
	);
	
	expect(container).toMatchSnapshot();
});