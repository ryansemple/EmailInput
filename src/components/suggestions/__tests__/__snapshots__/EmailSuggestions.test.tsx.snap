// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders EmailSuggestions component correctly. 1`] = `
<div>
  <ul
    class="EmailSuggestions full_width"
  />
</div>
`;
